src/jenkins.py

```diff
@@ -705,3 +708,196 @@ def get_agent_name(unit_name: str) -> str:
         The agent node name registered on Jenkins server.
     """
     return unit_name.replace("/", "-")
+
+
+PLUGIN_NAME_GROUP = r"^([a-zA-Z0-9-_]+)"
+WHITESPACE = r"\s*"
+VERSION_GROUP = r"\((.*?)\)"
+DEPENDENCIES_GROUP = r"\[(.*?)\]"
+PLUGIN_CAPTURE = rf"{PLUGIN_NAME_GROUP}{WHITESPACE}{VERSION_GROUP}"
+PLUGIN_LINE_CAPTURE = rf"{PLUGIN_CAPTURE} => {DEPENDENCIES_GROUP}"
+
+
+def _get_plugin_name(plugin_info: str) -> str:
+    """Get plugin name given a plugin info string of format <plugin-shortname> (<plugin-version>).
+
+    Args:
+        plugin_info: Text containing plugin name and plugin version.
+
+    Raises:
+        ValidationError: if the plugin info string does not conform to expected format.
+
+    Returns:
+        The plugin shortname.
+    """
+    match = re.match(PLUGIN_CAPTURE, plugin_info)
+    if not match:
+        raise ValidationError(f"No plugin matched in: {plugin_info}")
+    return match.group(1)
+
+
+def _build_dependencies_lookup(
+    plugin_dependency_outputs: typing.Iterable[str],
+) -> dict[str, tuple[str, ...]]:
+    """Build a lookup table of plugin short name to list of dependency plugin's short names.
+
+    Args:
+        plugin_dependency_outputs: The plugin dependency output from Jenkins Groovy script.
+
+    Returns:
+        The dependency lookup table.
+    """
+    dependency_lookup: dict[str, tuple[str, ...]] = {}
+    for line in plugin_dependency_outputs:
+        match = re.match(PLUGIN_LINE_CAPTURE, line)
+        if not match:
+            continue
+        plugin, dependencies = match.group(1), match.group(3)
+        if not dependencies:
+            dependency_lookup[plugin] = ()
```

Comment:
If using a defaultdict, you can remove this line

Reply:
I need to have the keys to build the set of all plugins, used by `plugins_lookup.keys()` later to subtract installed plugins set from all plugins set. :)

Reply:
The defaultdict can return an empty set by default, iirc. Am I wrong?

Reply:
It can, however, I still need the keys to be registered on the lookup table later on even if they are by default empty. For example, if I have
`{"a": [], "b": [], "c": []}`, the default dict implementation would be fine with `{}`. However, the keys information is lost. I later need the `plugins_lookup.keys()` which with default dict, would return `[]` if I don't register the keys.

Reply:
Noted ;)